fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/MetricUnionAvailability.java

```diff
@@ -0,0 +1,258 @@
+
+        // get a map from availability to its available metric columns intersected with configured metric columns
+        // i.e. metricColumns
+        availabilitiesToAvailableColumns = physicalTables.stream()
+                .collect(
+                        Collectors.toMap(
+                                PhysicalTable::getAvailability,
+                                physicalTable ->
+                                        Sets.intersection(
+                                                Utils.getSubsetByType(
+                                                        physicalTable
+                                                                .getAvailability()
+                                                                .getAllAvailableIntervals()
+                                                                .keySet(),
+                                                        MetricColumn.class
+                                                ),
+                                                metricColumns
+                                        )
+                        )
+                );
+
+        // validate metric uniqueness such that
+        // each table's underlying datasource schema don't have repeated metric column
+        Map<MetricColumn, Set<Availability>> duplicates = getDuplicateValue(availabilitiesToAvailableColumns);
+        if (!duplicates.isEmpty()) {
+            String message = String.format(
+                    "While constructing MetricUnionAvailability, Metric columns are not unique - %s",
+                    duplicates
+            );
+            LOG.error(message);
+            throw new IllegalArgumentException(message);
+        }
+
+        dataSourceNames = availabilitiesToAvailableColumns.keySet().stream()
+                .map(Availability::getDataSourceNames)
+                .flatMap(Set::stream)
+                .collect(
+                        Collectors.collectingAndThen(
+                                Collectors.toSet(),
+                                Collections::unmodifiableSet
+                        )
+                );
+    }
+
+    @Override
+    public Set<TableName> getDataSourceNames() {
+        return dataSourceNames;
+    }
+
+    /**
+     * Combines and returns intervals of all availabilities' metric columns.
+     * <p>
+     * Intervals of the same metric column are associated with the same metric column key. Overlapping intervals under
+     * the same metric column key are collapsed into single interval.
+     *
+     * @return a map of metric column to all of its available intervals in union
+     */
+    @Override
+    public Map<Column, List<Interval>> getAllAvailableIntervals() {
+        // get all availabilities
+        Set<Availability> availabilities = availabilitiesToAvailableColumns.keySet();
+
+        // take available interval maps from all availabilities and merge the maps together
+        return Stream.of(availabilities)
```

Comment:
:) It's possible. Though I may have been referring to splitting some other aspect of what this was doing previously. I often tend to pack more into 1 chained expression, rather than breaking things up, much to others' chagrin I'm sure 😉